### ISSUE TITLE
proxy complains about `no such image` when starting containers that require image downloading

### ISSUE THREAD

**Issue by oldlam** [weaveworks/weave#967]
while using weave proxy
for example if the docker host is clean and doesn't have an ubuntu image there
if i run `docker run -it ubuntu`
a message saying `Error response from daemon: no such image` is being returned
might be while facing this error, proxy could try to pull the image instead?

**Comment by rade** [comment]
I cannot reproduce this:
````
$ weave launch && weave launch-dns && weave launch-proxy
1a9b64b512bb8c02fa28d1e27d86f661ae4a493532422f743127a93b9eb3c6d4
5f7d8b6f7b10a32f2a9307f061113ccad6c1dffcfb27b402f8fc0ccb3fdda352
83d095e1241a6828fe821813aeefbccb6dbb9e707dfbbbb06f7811033af5e5e7
vagrant@host1:~$ eval $(weave proxy-env)
vagrant@host1:~$ weave version
weave script 1.0.0
weave router 1.0.0
weave DNS 1.0.0
weave exec 1.0.0 
vagrant@host1:~$ docker run -ti ubuntu
Unable to find image 'ubuntu:latest' locally
latest: Pulling from ubuntu
428b411c28f0: Pull complete 
435050075b3f: Pull complete 
9fd3c8c9af32: Pull complete 
6d4946999d4f: Already exists 
ubuntu:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.
Digest: sha256:45e42b43f2ff4850dcf52960ee89c21cda79ec657302d36faaaa07d880215dd9
Status: Downloaded newer image for ubuntu:latest
root@92fd2b50bab4:/# 
````
What are you doing differently? Are you running a version earlier than 1.0? If so then please upgrade.

**Comment by oldlam** [comment]
I'm basically doing same stuff, and I just tried to run the exact same steps as u do.
```
$ weave launch && weave launch-dns && weave launch-proxy
0ba180ca324e5939f7633106f052b30a6dfe6531485192cadec435711cfbfc75
5cbd9070431122fc03de82d2c28b24918d3885763c65b02e007f619de34381be
e2117271ec8c6f767c0d69b8240f3a91f521ca7566fabc77ddb06134eef8fdc8
$ eval $(weave proxy-env)
$ weave version
weave script 1.0.0
weave router 1.0.0
weave DNS 1.0.0
weave exec 1.0.0
$ docker run -ti ubuntu
Error response from daemon: no such image
```
one more detail is that my docker version is 1.7 (which is in digital ocean):
```
$ docker version
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```

**Comment by rade** [comment]
ouch! looks like something changed/broke in docker 1.7.0. I was running 1.6.2 but am getting the same result as you with 1.7.0.

Thanks for the bug report!

**Comment by oldlam** [comment]
You're so welcome and so much appreciated for the quick responses!!
Cheers! :+1:

**Comment by rade** [comment]
This is caught be one of our smoke tests...
````
$ ./610_proxy_wait_for_weave_test.sh 
Proxy waits for weave to be ready before running container commands
da6994a4cd0cf5ab6f764c0ec5d4f60ea3cf9183830d806c297796bb5b09eef3
Error response from daemon: No such image: busybox:latest
Error: failed to remove images: [busybox]
````

**Comment by rade** [comment]
I found [the docker commit that broke this](https://github.com/docker/docker/commit/013fb87543a47524efba3c371a1d22afe151287c)

Previously, any 404 return from the server would cause the client to pull the image. Now only 404 returns mentioning the image names do. A 404 for a missing image in `/containers/create` *does* mention the image name, which is why things still work in docker land. But the proxy does an `inspect` on the image, and the 404 on that does *not* mention the image name. The proxy returns that error to the client. Hence the breakage.

**Comment by rade** [comment]
> But the proxy does an inspect on the image, and the 404 on that does *not* mention the image name. 

Actually, it does, but it's [swallowed by the docker client library](https://github.com/fsouza/go-dockerclient/blob/master/image.go#L179) we use in the proxy.

**Comment by rade** [comment]
@oldlam I've just released [weave 1.0.1](https://github.com/weaveworks/weave/releases/tag/v1.0.1) with a fix. Thanks again for reporting this problem.

**Comment by oldlam** [comment]
@rade thanks so much! you guys are awesome, support!
